src/main/java/uk/gov/pay/products/service/ProductFinder.java

```diff
@@ -92,26 +95,31 @@ public Boolean deleteByGatewayAccountIdAndExternalId(Integer gatewayAccountId, S
         return productDao.findByGatewayAccountId(gatewayAccountId)
                 .stream()
                 .map(ProductEntity::toProduct)
-                .collect(Collectors.toList())
+                .map(linksDecorator::decorate)
+                .collect(Collectors.toUnmodifiableList());
+    }
+
+    @Transactional
+    public List<Product> findByGatewayAccountIdAndType(Integer gatewayAccountId, ProductType type) {
+        return productDao.findByGatewayAccountIdAndType(gatewayAccountId, type)
                 .stream()
+                .map(ProductEntity::toProduct)
                 .map(linksDecorator::decorate)
-                .collect(Collectors.toList());
+                .collect(Collectors.toUnmodifiableList());
     }
 
     @Transactional
     public Optional<Product> findByProductPath(String serviceNamePath, String productNamePath) {
         return productDao.findByProductPath(serviceNamePath, productNamePath)
-                .map(productEntity -> linksDecorator.decorate(productEntity.toProduct()));
+                .map(ProductEntity::toProduct)
+                .map(linksDecorator::decorate);
     }
 
     @Transactional
     public List<ProductUsageStat> findProductsAndUsage(Integer gatewayAccountId) {
         return productDao.findProductsAndUsage(gatewayAccountId)
                 .stream()
-                .map(productUsageStat -> {
-                    productUsageStat.setProduct(linksDecorator.decorate(productUsageStat.getProduct()));
-                    return productUsageStat;
-                })
-                .collect(Collectors.toList());
+                .peek(productUsageStat -> productUsageStat.setProduct(linksDecorator.decorate(productUsageStat.getProduct())))
```

Comment:
nice

Reply:
You can thank IntelliJ’s suggestions for that!